retryablemessageprocessor.go

```diff
@@ -0,0 +1,74 @@
+package messageprocessor
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/rs/xstats"
+)
+
+// MaxRetries is the maximum number of time's we'd like to retry processing a record before quitting
+const consumerRetriesExceeded = "kinesis.consumer_error.retries_exceeded"
+
+// RetryableMessageProcessor is a `MessageProcessor` decorator that re-attempts
+// processing of messages 'maxAttempts' number of times in case of failures
+// 'maxAttempts' is a configurable parameter which can be set by consumer of this lib
+// Exponential backoff has been implemented as a retry mechanism
+type RetryableMessageProcessor struct {
+	maxAttempts int
+	wrapped     MessageProcessor
+}
+
+// ProcessMessage invokes the wrapped `MessageProcessor`. Attempts retries using exponential backoff
+// if underlying 'MessageProcessor' returns an error.
+// If 'maxAttempts' are exceeded without successful processing, it emits a stat indicating the same
+func (t *RetryableMessageProcessor) ProcessMessage(ctx context.Context, record *kinesis.Record) MessageProcessError {
+	stat := xstats.FromContext(ctx)
+	var messageProcErr MessageProcessError
+	var attemptNum int
+	for attemptNum < t.maxAttempts {
+		messageProcErr = t.wrapped.ProcessMessage(ctx, record)
+		if messageProcErr != nil && messageProcErr.Error() != nil && messageProcErr.IsRetryable() {
+			if messageProcErr.RetryAfter() > 0 {
+				// Wait for duration specified in 'Retry-After'
+				waitRetryAfter(messageProcErr.RetryAfter())
+			} else {
+				// Or perform exponential backoff
+				waitToRetry(attemptNum)
+			}
+		} else {
+			break
+		}
+		attemptNum++
+	}
+
+	if attemptNum >= t.maxAttempts {
+		stat.Count(consumerRetriesExceeded, 1)
```

Comment:
suggestion: If the conditional on line 47 is met, I think we should wrap the existing error in a new error type named something like `MaxRetriesExceededError`, whose `RetryAfter` method will return `false`, and whose `Error` method wraps prepends something like "<t.maxAttempts> retries exceeded: " to the original error message from its wrapped error instance.

This will give the application code or any outer decorators more information on how to handle the error returned here, should they choose to use it.

Reply:
Added support for `MaxRetriesExceededError`